lib/membrane_http_adaptive_stream/sink.ex

```diff
@@ -241,6 +326,108 @@ defmodule Membrane.HTTPAdaptiveStream.Sink do
     end
   end
 
+  # we are operating on partial segments which may require
+  # assembling previous partial segments and creating regular one
+  # before processing the current segment
+  defp handle_partial_segment(
+         buffer,
+         track_id,
+         segment_duration,
+         %{mode: :live} = state
+       ) do
+    %{
+      duration: duration,
+      independent?: independent?
+    } = buffer.metadata
+
+    partial_segments = get_in(state, [:track_to_partial_segments, track_id]) || []
+    partial_segments_duration = Enum.reduce(partial_segments, 0, &(&1.metadata.duration + &2))
+
+    {full_segment, changeset, manifest, partial_segments} =
+      if independent? and duration + partial_segments_duration >= segment_duration.min and
+           partial_segments != [] do
+        # create a full segment by merging all pending partial segments
+        payload = %Membrane.Buffer{
+          payload:
+            partial_segments |> Enum.map(& &1.payload) |> Enum.reverse() |> IO.iodata_to_binary(),
+          metadata: %{
+            duration: partial_segments_duration
+          }
+        }
+
+        {changeset, manifest} = Manifest.finalize_last_segment(state.manifest, track_id)
+
+        {payload, changeset, manifest, []}
+      else
+        {nil, nil, state.manifest, partial_segments}
+      end
+
+    new_segment_necessary? = Enum.empty?(partial_segments)
+
+    bytes_offset = Enum.reduce(partial_segments, 0, &(byte_size(&1.payload) + &2))
+    partial_segments = [buffer | partial_segments]
+
+    state = put_in(state, [:track_to_partial_segments, track_id], partial_segments)
+
+    manifest =
+      if new_segment_necessary? do
+        {_changeset, manifest} =
+          Manifest.add_segment(
+            manifest,
+            track_id,
+            0,
+            0,
+            [{:partial?, true} | creation_time(state.mode)]
+          )
+
+        manifest
+      else
+        manifest
+      end
+
+    byte_range = {byte_size(buffer.payload), bytes_offset}
+
+    buffer = %Membrane.Buffer{
+      buffer
+      | metadata: Map.put(buffer.metadata, :byte_range, byte_range)
+    }
+
+    manifest
+    |> Manifest.add_partial_segment(
+      track_id,
+      independent?,
+      duration,
+      byte_range: byte_range,
+      independent: independent?
+    )
+    |> then(fn {new_changeset, manifest} ->
+      changesets = reject_nils([changeset, new_changeset])
+      segments = reject_nils([full_segment, buffer])
```

Comment:
this is quite confusing, especially since these lists are zipped in `handle_write`